components/suggest/src/weather.rs

```diff
@@ -1,53 +1,235 @@
+    max_keyword_word_count: usize,
+}
+
 impl SuggestDao<'_> {
     /// Fetches weather suggestions.
     pub fn fetch_weather_suggestions(&self, query: &SuggestionQuery) -> Result<Vec<Suggestion>> {
-        // Weather keywords are matched by prefix but the query must be at least
-        // three chars long. Unlike the prefix matching of other suggestion
-        // types, the query doesn't need to contain the first full word.
+        // We'll just stipulate we won't support tiny queries in order to avoid
+        // a bunch of work when the user starts typing a query.
         if query.keyword.len() < 3 {
             return Ok(vec![]);
         }
 
-        let keyword_lowercased = &query.keyword.trim().to_lowercase();
-        let suggestions = self.conn.query_rows_and_then_cached(
+        // The first step in parsing the query is lowercasing and splitting it
+        // into words. We want to avoid that work for strings that are so long
+        // they can't possibly match. The longest possible weather query is two
+        // geonames + one weather keyword + at least two spaces between those
+        // three components, say, 10 spaces total for some wiggle room. There's
+        // not much point in an analogous min length check since weather
+        // suggestions can be matched on city alone and many city names are only
+        // a few characters long ("nyc").
+        let g_cache = self.geoname_cache();
+        let w_cache = self.weather_cache();
+        let max_query_len = 2 * g_cache.max_name_length + w_cache.max_keyword_length + 10;
+        if max_query_len < query.keyword.len() {
+            return Ok(vec![]);
+        }
+
+        let max_win_size =
+            std::cmp::max(g_cache.max_name_word_count, w_cache.max_keyword_word_count);
+        let kw = query.keyword.to_lowercase();
+        let paths = filter_map_chunks::<Token>(&kw, max_win_size, |chunk, chunk_index, path| {
+            // Match the chunk to a token type that hasn't already been matched
+            // in this path.
+            for tt in [
+                TokenType::City,
+                TokenType::Region,
+                TokenType::WeatherKeyword,
+            ] {
+                if !path.iter().any(|t| t.token_type() == tt) {
+                    let tokens = self.match_weather_tokens(tt, path, chunk, chunk_index == 0)?;
+                    if !tokens.is_empty() {
+                        // This chunk matches, so return the tokens to continue
+                        // down this path.
+                        return Ok(Some(tokens));
+                    }
+                }
+            }
+            // This chunk doesn't match. Return `None` to filter out the path.
+            Ok(None)
+        })?;
+
+        // `paths` contains all matching paths, and if a path has both a city
+        // and region, the city is in the region since we filtered out city-
+        // region mismatches above via `match_weather_tokens()`.
```

Comment:
Okay, I do have one suggestion.  What if the `filter_map_chunks` only allowed the mapper to return an `Option` rather than a `Vec`?  I think that makes that function much simpler to describe and understand, although though the implementation basically stays the same.

That pushes the complexity out to this code, but I don't think that's too hard to handle, we just need a second dose of recursion.  The rest of this function could be replaced with something like this:
```
    // no changes before here
    let mut suggestions = vec![];
    collect_suggestions_recurse(&mut suggestions, &paths, &mut Vec::new());
    Ok(suggestions);
}

fn collect_suggestions_recurse(
    suggestions: &mut Vec<Suggestion>,
    remaining: &[Token],
    current: &mut Vec<Token>)
{
    // check if the previous code generated an mismatched series of tokens (City:NewYorkCity +
    // Region:CA)
    if tokens_mismatched(tokens) {
        return;
    }
    match remaining.split_first() {
        None => {
            if let Some(s) = tokens_to_suggestion(current) {
                suggestions.push(s)
            }
        }
        Some((first, remaining)) => {
            current.push(first);
            collect_suggestions_recurse(suggestions, remaining, current);
            current.pop();
            // Token::alternate_tokens generates a Vec of other ways to interpret the token.  For
            // example, when we parse "New York" in `filter_map_chunks` we would parse it as
            // `Token::City("New York)`, but then alternative_tokens would generate
            // `Token::Region("New York")`.  If there's only 1 possible alternate, then it
            // could return an option instead of a vec.
            for alt in first.alternate_tokens() {
                current.push(alt);
                collect_suggestions_recurse(suggestions, remaining, current);
                current.pop();
            }
        }
    }
}
```


I think my main issue conceptually is there are 2 sources of branching: different chunk combinations and the different ways to parse strings into tokens.  IMO, it would be simpler if those 2 things were split out into separate processes.

Reply:
However, I really meant it when I said don't block this PR on this issue.  If you disagree then feel free to merge first and we can discuss if this refactoring makes sense or not in a separate one.